### DISCUSSION TITLE
Using Kotlin in the Android app

### DISCUSSION THREAD

**Discussion by arnaudvergnet** [organicmaps/organicmaps#2212] [Q&A] [answered]
Maybe this has already been brought up, but I could not find information in the repo.

Since Java and Kotlin are interoperable (see [this](https://kotlinlang.org/docs/java-interop.html) and [this](https://kotlinlang.org/docs/java-to-kotlin-interop.html)) I was wondering if you would accept new contributions in Kotlin.

Kotlin is the official language for Android and new frameworks are written for it (Jetpack Compose for example), this would make the app easier to maintain in the long run.

**Comment by biodranik** [comment, marked as answer]
Kotlin is good for development and for using 3party libraries. But it is worse for performance and device's battery: https://medium.com/rsq-technologies/comparative-evaluation-of-selected-constructs-in-java-and-kotlin-part-1-dynamic-metrics-2592820ce80

If you see any significant blockers/reasons why we should start using Kotlin now, please share, and let's discuss it.

Regarding 3party libraries, what is really required/needed/missing for OM in your opinion?

**Reply by andrewshadura** [reply]
I'm not sure why performance is important for the front-end given that all CPU-heavy code is in C++. The performance hit is likely going to be negligible compared to the huge safety and convenience improvements.

**Reply by biodranik** [reply]
KISS + YAGNI.

**Comment by arnaudvergnet** [comment]
I did not know about the performance difference! Maybe this could be mitigated by keeping key parts in java. But then this would make the codebase harder to read as it would mix 2 languages.

I did not find any blockers to start using Kotlin, but I tried it and felt it was easier to work with it. Less boilerplate code, more null safety and other benefits for developers. As I was looking into doing some refactoring on the Android app, I was wondering if it would be a good idea to write new classes in Kotlin.

Regarding 3party libraries, I saw Jetpack Compose was Kotlin only, so I felt more libraries could start following this trend. This could one day prevent OM to use such libraries.

**Reply by biodranik** [reply]
IMO now it would be better to focus on existing bugs and features. And switch to Kotlin only when we hit a real blocker. What do you think?

**Reply by arnaudvergnet** [reply]
Yes that would be the best solution for now, thanks for answering my questions!

**Comment by andrewshadura** [comment]
In this case, Kotlin is KISS, as it drastically reduces the complexity of the code.

On Mon, 14 Mar 2022, at 14:44, Alexander Borsuk wrote:
> 
> 
> KISS + YAGNI.
> 

-- 
Cheers,
  Andrej

**Reply by biodranik** [reply]
Migrating existing code to Kotlin is not KISS.

**Comment by andrewshadura** [comment]
I don't see why given that there are people willing to do the work. OTOH there are a lot of people who won't even touch Java code with a ten foot pole, but would be willing to work on Kotlin code (myself included), so it may actually bring new contributors.

On Mon, 14 Mar 2022, at 14:52, Alexander Borsuk wrote:
> 
> 
> Migrating existing code to Kotlin is not KISS.
> 
> 
> —
> 

-- 
Cheers,
  Andrej

**Reply by biodranik** [reply]
A good Android programmer usually reads and writes Java code too. We consider switching when it becomes really necessary, or benefits obviously overweight cons (time to refactor, runtime and battery overhead).

**Reply by biodranik** [reply]
I would like to add that the main issue now is not with the Java itself, but with many overcomplicated patterns used to implement simple features. Android and iOS implementations can and should be made simpler first. Repeating the same complex patterns in Kotlin doesn't make any sense.

**Comment by arnaudvergnet** [comment]
I'm back with some more refactoring ideas! I've been looking a bit at [Jetpack Compose](https://developer.android.com/jetpack/compose), and I think it would be worth considering.

It is no secret that the current Android app is a mess, especially if you look at the layout folder. Compose will help solve that issue by writing reusable components in Kotlin. Their approach at building UI looks a lot like QML or React.

Migration can be done incrementally as explained [here](https://developer.android.com/codelabs/jetpack-compose-migration). We could start with simple screens such as the about page, and as we write reusable components the rest of the screens would be easier. It would also help in migrating to Material 3 theme as discussed in https://github.com/organicmaps/organicmaps/discussions/4676. We could then make 3 migrations at once: Kotlin, Compose and Material 3.

Google seems to be really pushing this framework forward and it seems they want to make it the official method for building apps. If there is no noticeable impact on performance, I really think it would be nice to start working with it instead of waiting to be limited by the Java/XML views. While refactoring over-complicated patterns, why not migrate to Compose at the same time? It would save a lot of development time.

It is even possible to write [cross platform apps for desktop](https://www.jetbrains.com/lp/compose-mpp/). As desktop do not require as much optimization as mobile devices, we could share the android code with the desktop version so it is more up to date. 

A Compose app seems to be less complex and easier to maintain than a regular XML views app. This could help on-boarding new developers on the project. And as it is the new trendy Android tech stack (v1 was released in July 2021), it could help find new potential contributors.

Keep in mind I have no experience on Compose yet, so everything I say here is not out of experience, but from information I found online. We need to lookup the [limitations](https://developer.android.com/jetpack/androidx/compose-roadmap), the impact on performance (especially when mixing XML and Compose), and ask experienced Compose devs for their thoughts on such migration.

**Reply by arnaudvergnet** [reply]
@biodranik @rtsisyk @vng any thoughts on this? Do any of you have experience with Compose?

**Reply by biodranik** [reply]
1. QML and React both suck in terms of performance and battery life.
2. Google is well-known for its history of pushing and then closing/terminating its products.
3. OM is now significantly faster and less battery consuming than many competitors. IMO, avoiding unnecessary fancy frameworks contributes to these "selling features" that are too important to trade for "potential new contributors who do not know how Java works under the hood".

**Reply by arnaudvergnet** [reply]
> QML and React both suck in terms of performance and battery life.

Yes because these are not Android native technologies and use a lot of abstraction. I am only talking about the way to build UI in declarative style. The performance should be about the same as a Java/XML application because it is still an Android native stack.

> Google is well-known for its history of pushing and then closing/terminating its products.

True and that is quite annoying when choosing technologies. In that case Compose and Kotlin seem to be here to stay. When you look over the Android documentation, Kotlin/Compose examples are starting to be all over the place.

> OM is now significantly faster and less battery consuming than many competitors. IMO, avoiding unnecessary fancy frameworks contributes to these "selling features" that are too important to trade for "potential new contributors who do not know how Java works under the hood".

The performance aspect is why I think it could be worth it to try and run some tests. And I'm not talking only about attracting new developers, but also making the current code base simpler. Declarative style UI used in Compose is quite common now in other frameworks. And from my experience with such other frameworks, building UI this way is a lot simpler than the current XML approach.

For example, when you want to build an adaptive layout in XML you have to create a mess of files to handle all the edge cases and it becomes hard to fix bugs. [Using Compose](https://developer.android.com/jetpack/compose/layouts/adaptive) it seems a lot simpler, you keep the logic inside the component.

I'm only saying if we try it and it does not impact performance too much, then we could use Compose for upcoming refactorings to gradually make the switch. OM could really benefit from this in the long term IMO.

**Reply by biodranik** [reply]
We can check hypotheses by:
1. Enabling kotlin/compose frameworks and seeing their apk overhead first.
2. Choosing some existing complex layout (About dialog?) and comparing how easier it would be to support it with a new approach.
3. Compare its runtime performance on some old Android 5 device.

Then we can have more data to decide.

**Reply by arnaudvergnet** [reply]
Yes this was exactly the idea! When I have some time I'll try do make some tests

**Reply by arnaudvergnet** [reply]
@biodranik I did some initial testing for the apk size. I simply added the dependencies and replaced the help screen with a compose view and here are the results when compiling with the fdroidRelease build variant:

**Without Compose**:
- apk: 57.6MB
- Installed app: 79.10MB

**With Compose**
- apk: 60.2MB
- Installed app: 90.21MB

But on my phone the FDroid version is already 90.49MB, so there does not seem to be a big difference. Especially if we take into account the size of the maps. For the apk size, the google play variant uses app bundles so it should do further optimizations.

**Reply by biodranik** [reply]
It would be great to measure the battery usage somehow "honestly".

**Reply by arnaudvergnet** [reply]
The only place where it's going to make a difference is for views over the map, the rest will have a meaningless impact. So we'll only be able to measure this when we fully migrate the views on the map (button, place page, ...).

Where there could be a difference is for the loading of pages (see this comparison here https://medium.com/okcredit/comparing-jetpack-compose-performance-with-xml-9462a1282c6b), but there are ways to optimize this.

Compose is still a native technology, so I don't think it will impact performance that much. I've used some apps that did the switch recently and I did not notice any regression in performance.

Guess I'll try to further migrate the about screen to see if we can find any issues.

**Reply by bam80** [reply]
> Compose will help solve that issue by writing reusable components in Kotlin. Their approach at building UI looks a lot like QML or React.

FYI:
we already started to implement QML/Qt Quick/Kirigami UI as GSoC project.
The current plan is to implement convergent UI for Linux Mobile platforms and Desktop.
But that should be portable to Android and all the other platforms too, at least as "experimental" UI.

**Reply by arnaudvergnet** [reply]
Nice that you use kirigami, it will work well on mobile linux. But on android Qt is still very bad. I've used a few apps from KDE using Qt on Android and none of them have a native look and feel, and most of them are slow. Maybe in the future if QT has better support for Android we could share the same codebase, but for now this is not possible.

**Reply by bam80** [reply]
Yeah, common codebase would be ideal. At least we could keep it as experimental UI I hope.
We faced an issue with the scroll speed of one Kirigami app on Android wich is apparently solvable, but other than that I'm not aware any slowness. Do you have more details with the references?

**Reply by biodranik** [reply]
Unfortunately, the common UI codebase for different platforms newer was ideal, in decades. Writing native platform code without any "cross-platform" overhead produces better quality software, that looks natively and consumes less battery. These are very important features that should be always kept in mind.

**Reply by bam80** [reply]
It probably deserve separate discussion but in my understanding in QML case there is just no any "cross-platform" overhead by design - the apps speak to hardware directly via OpenGL or other native graphics API, so there are just no any translation layers capable to provoke the overhead.
But that is theory, I don't know how it will be on practice. And that's what I would like to know.

**Reply by arnaudvergnet** [reply]
For cross platform stuff, see https://github.com/orgs/organicmaps/discussions/1865.

Cross-platform UI only works if you have a very specific design and you don't care about respecting the platform's native look. QML apps look really out of place and don't respect Android's gestures or styling. If you want to respect the platform's styling/look and feel it will require more work and make the code more complex. Qt works really great for convergent applications, but cross-platform is another issue.

Regarding performance, I haven't done any proper testing, its just the feeling I got after using some apps (might be because the gestures are not responding the same way as native android apps). Qt on Android is still in its early days so it can probably be improved. And Qt apps have to ship their own rendering engine if they don't use native components so it will probably increase the app size.

**Reply by bam80** [reply]
> Cross-platform UI only works if you have a very specific design and you don't care about respecting the platform's native look. QML apps look really out of place and don't respect Android's gestures or styling. If you want to respect the platform's styling/look and feel it will require more work and make the code more complex. Qt works really great for convergent applications, but cross-platform is another issue.

From what I can see, that's not quite true, at least in theory - Qt Quick Controls support [_styles_](https://doc.qt.io/qt-6/qtquickcontrols-styles.html), and [Material Style](https://doc.qt.io/qt-6/qtquickcontrols-material.html) based on the Google Material Design Guidelines is one of those.

**Reply by bam80** [reply]
There is interesting comparison: 
[Flutter vs Qt QML for cross-platform app development](https://pkoretic.medium.com/flutter-vs-qt-qml-9a2d655b5ab1)

PS:
oh, just noted you don't speak about Flutter here..

**Comment by biodranik** [comment]
Do you have any information on compose overhead for the battery life? I don't believe that this "code simplicity" comes for free. For example, if there is a choice of using Java instead of Kotlin, but saving one hour of battery life for a day-long trip, then the answer is obviously Java. We develop apps for users in the first place.

Looks like there are some [optimization techniques already for compose](https://medium.com/icerock/optimize-or-die-profiling-and-optimization-in-jetpack-compose-a165c8897b3f). Does it mean that we can hit similar issues?

**Comment by biodranik** [comment]
Let's reconsider Kotlin in its minimal form, without overhead, to make the code more readable/less verbose: https://github.com/organicmaps/organicmaps/pull/12588
